internal/provider/kubernetes/tlsroute.go

```diff
@@ -0,0 +1,326 @@
+		route := routeList.Items[i]
+		routeKey := utils.NamespacedName(&route)
+		if routeKey == request.NamespacedName {
+			found = true
+		}
+
+		// Store the tlsroute in the resource map.
+		r.resources.TLSRoutes.Store(routeKey, &route)
+		log.Info("added tlsroute to resource map")
+
+		// Get the route's namespace from the cache.
+		nsKey := types.NamespacedName{Name: route.Namespace}
+		ns := new(corev1.Namespace)
+		if err := r.client.Get(ctx, nsKey, ns); err != nil {
+			if errors.IsNotFound(err) {
+				// The route's namespace doesn't exist in the cache, so remove it from
+				// the namespace resource map if it exists.
+				if _, ok := r.resources.Namespaces.Load(nsKey.Name); ok {
+					r.resources.Namespaces.Delete(nsKey.Name)
+					log.Info("deleted namespace from resource map")
+				}
+			}
+			return reconcile.Result{}, fmt.Errorf("failed to get namespace %s", nsKey.Name)
+		}
+
+		// The route's namespace exists, so add it to the resource map.
+		r.resources.Namespaces.Store(nsKey.Name, ns)
+		log.Info("added namespace to resource map")
+
+		// Get the route's backendRefs from the cache. Note that a Service is the
+		// only supported kind.
+		for i := range route.Spec.Rules {
+			for j := range route.Spec.Rules[i].BackendRefs {
+				ref := route.Spec.Rules[i].BackendRefs[j]
+				if err := validateTLSRouteBackendRef(&ref); err != nil {
+					return reconcile.Result{}, fmt.Errorf("invalid backendRef: %w", err)
+				}
+
+				// The backendRef is valid, so get the referenced service from the cache.
+				svcKey := types.NamespacedName{Namespace: route.Namespace, Name: string(ref.Name)}
+				svc := new(corev1.Service)
+				if err := r.client.Get(ctx, svcKey, svc); err != nil {
+					if errors.IsNotFound(err) {
+						// The ref's service doesn't exist in the cache, so remove it from
+						// the resource map if it exists.
+						if _, ok := r.resources.Services.Load(svcKey); ok {
+							r.resources.Services.Delete(svcKey)
+							log.Info("deleted service from resource map")
+						}
+					}
+					return reconcile.Result{}, fmt.Errorf("failed to get service %s/%s",
+						svcKey.Namespace, svcKey.Name)
+				}
+
+				// The backendRef Service exists, so add it to the resource map.
+				r.resources.Services.Store(svcKey, svc)
+				log.Info("added service to resource map")
+			}
+		}
+	}
+
+	if !found {
+		// Delete the tlsroute from the resource map.
+		r.resources.TLSRoutes.Delete(request.NamespacedName)
+		log.Info("deleted tlsroute from resource map")
+
+		// Delete the Namespace and Service from the resource maps if no other
+		// routes (TLSRoute or HTTPRoute) exist in the namespace.
+		found, err := isRoutePresentInNamespace(ctx, r.client, request.NamespacedName.Namespace)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if !found {
+			r.resources.Namespaces.Delete(request.Namespace)
+			log.Info("deleted namespace from resource map")
+			r.resources.Services.Delete(request.NamespacedName)
```

Comment:
This attempts to delete a possible non-existent Service, and not the relevant ones. It is introducing an issue in the tlsroute controller, which was also present for httproute controller too. I have filed an issue for this https://github.com/envoyproxy/gateway/issues/536